crates/forge/src/test_case_summary.rs

```diff
@@ -46,24 +71,78 @@ impl TestCaseSummary {
 }
 
 #[must_use]
-pub fn extract_result_data(run_result: &RunResult) -> Option<String> {
-    let data = match &run_result.value {
-        RunResultValue::Panic(data) | RunResultValue::Success(data) => data,
-    };
+/// Returns a string with the data that was produced by the test case.
+/// If the test case was successful, it returns the data that was produced by the test case.
+/// If the test case failed, it returns a string comparing the panic data and the expected data.
+pub fn extract_result_data(
+    run_result: &RunResult,
+    expectation: &TestExpectation,
+) -> Option<String> {
+    match &run_result.value {
+        RunResultValue::Success(data) => {
+            let mut readable_text = String::new();
 
-    let mut readable_text = String::new();
+            for felt in data {
+                readable_text.push_str(&format!("\n    original value: [{felt}]"));
+                if let Some(short_string) = as_cairo_short_string(felt) {
+                    readable_text.push_str(&format!(", converted to a string: [{short_string}]"));
+                }
+            }
 
-    for felt in data {
-        readable_text.push_str(&format!("\n    original value: [{felt}]"));
-        if let Some(short_string) = as_cairo_short_string(felt) {
-            readable_text.push_str(&format!(", converted to a string: [{short_string}]"));
+            if readable_text.is_empty() {
+                None
+            } else {
+                readable_text.push('\n');
+                Some(readable_text)
+            }
         }
-    }
+        RunResultValue::Panic(panic_data) => {
+            let expected_data = match expectation {
+                TestExpectation::Panics(panic_expectation) => match panic_expectation {
+                    PanicExpectation::Exact(data) => Some(data),
+                    PanicExpectation::Any => None,
+                },
+                TestExpectation::Success => None,
+            };
+
+            let panic_string: Vec<String> = panic_data
+                .iter()
+                .map(|felt| as_cairo_short_string(felt).unwrap_or_default())
+                .collect();
+
+            match expected_data {
+                Some(expected) if expected == panic_data => None,
+                Some(expected) => {
+                    let expected_string: Vec<String> = expected
+                        .iter()
+                        .map(|felt| as_cairo_short_string(felt).unwrap_or_default())
+                        .collect();
 
-    if readable_text.is_empty() {
-        None
-    } else {
-        readable_text.push('\n');
-        Some(readable_text)
+                    Some(format!(
+                        "\x1B[31mFAIL: Test did not meet expectations!\x1B[0m\n\
+  \x1B[32m  Actual:   {panic_data:?} ({panic_string:?})\x1B[0m\n\
+  \x1B[31m  Expected: {expected:?} ({expected_string:?})\x1B[0m\n"
```

Comment:
```suggestion
                    Some(format!(
                        "\nIncorrect panic data\n\
            Actual:   {panic_data:?} {panic_string:?}\n\
            Expected: {expected:?} {expected_string:?}\n"
```

I would also not use debug printing if possible for strings, instead just print vector elements one by one (it creates weird symbols like `\u`)

Reply:
I changed this a bit to keep the colored console output. Can you check once again?

Reply:
Ok, I will revert it, I thought it was a good idea because it's something I often saw in other test runners when comparing EXPECTED and ACTUAL